src/mito2/src/engine/truncate_test.rs

```diff
@@ -0,0 +1,441 @@
+    assert_eq!(expected, batches.pretty_print().unwrap());
+
+    // Truncate the region.
+    engine
+        .handle_request(region_id, RegionRequest::Truncate(RegionTruncateRequest {}))
+        .await
+        .unwrap();
+
+    // Put data to the region again.
+    let rows = Rows {
+        schema: column_schemas,
+        rows: build_rows(5, 8),
+    };
+    put_rows(&engine, region_id, rows).await;
+
+    // Scan the region.
+    let request = ScanRequest::default();
+    let stream = engine.handle_query(region_id, request).await.unwrap();
+    let batches = RecordBatches::try_collect(stream).await.unwrap();
+    let expected = "\
++-------+---------+---------------------+
+| tag_0 | field_0 | ts                  |
++-------+---------+---------------------+
+| 5     | 5.0     | 1970-01-01T00:00:05 |
+| 6     | 6.0     | 1970-01-01T00:00:06 |
+| 7     | 7.0     | 1970-01-01T00:00:07 |
++-------+---------+---------------------+";
+    assert_eq!(expected, batches.pretty_print().unwrap());
+}
+
+#[tokio::test]
+async fn test_engine_truncate_after_flush() {
+    let mut env = TestEnv::with_prefix("truncate-flush");
+    let engine = env.create_engine(MitoConfig::default()).await;
+
+    // Create the region.
+    let region_id = RegionId::new(1, 1);
+    let request = CreateRequestBuilder::new().build();
+    let column_schemas = rows_schema(&request);
+    engine
+        .handle_request(region_id, RegionRequest::Create(request))
+        .await
+        .unwrap();
+
+    // Put data to the region.
+    let rows = Rows {
+        schema: column_schemas.clone(),
+        rows: build_rows(0, 3),
+    };
+    put_rows(&engine, region_id, rows).await;
+
+    // Flush the region.
+    engine
+        .handle_request(region_id, RegionRequest::Flush(RegionFlushRequest {}))
+        .await
+        .unwrap();
+
+    let request = ScanRequest::default();
+    let scanner = engine.scan(region_id, request.clone()).unwrap();
+    assert_eq!(1, scanner.num_files());
+
+    // Truncate the region.
+    engine
+        .handle_request(region_id, RegionRequest::Truncate(RegionTruncateRequest {}))
+        .await
+        .unwrap();
+
+    // Put data to the region.
+    let rows = Rows {
+        schema: column_schemas,
+        rows: build_rows(5, 8),
+    };
+    put_rows(&engine, region_id, rows).await;
+
+    // Scan the region.
+    let stream = engine
+        .handle_query(region_id, request.clone())
+        .await
+        .unwrap();
+    let batches = RecordBatches::try_collect(stream).await.unwrap();
+    let expected = "\
++-------+---------+---------------------+
+| tag_0 | field_0 | ts                  |
++-------+---------+---------------------+
+| 5     | 5.0     | 1970-01-01T00:00:05 |
+| 6     | 6.0     | 1970-01-01T00:00:06 |
+| 7     | 7.0     | 1970-01-01T00:00:07 |
++-------+---------+---------------------+";
+    assert_eq!(expected, batches.pretty_print().unwrap());
+
+    tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
suggestion: Don't rely on sleep in tests if possible. You could use the listener to block until the flush is finished.
https://github.com/GreptimeTeam/greptimedb/blob/50cea29b29a45073a11b1cb3702a5eaa21b465cb/src/mito2/src/engine/flush_test.rs#L109-L110